fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneQueryClause.java

```diff
@@ -23,21 +23,180 @@
 import com.apple.foundationdb.annotation.API;
 import com.apple.foundationdb.record.EvaluationContext;
 import com.apple.foundationdb.record.PlanHashable;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.lucene.query.BitSetQuery;
 import com.apple.foundationdb.record.metadata.Index;
 import com.apple.foundationdb.record.provider.foundationdb.FDBRecordStoreBase;
 import com.apple.foundationdb.record.query.plan.cascades.explain.Attribute;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.BoostQuery;
+import org.apache.lucene.search.MultiPhraseQuery;
+import org.apache.lucene.search.PhraseQuery;
+import org.apache.lucene.search.PrefixQuery;
 import org.apache.lucene.search.Query;
+import org.apache.lucene.search.SynonymQuery;
+import org.apache.lucene.search.TermQuery;
+import org.apache.lucene.search.spans.SpanNearQuery;
+import org.apache.lucene.search.spans.SpanOrQuery;
+import org.apache.lucene.search.spans.SpanQuery;
+import org.apache.lucene.search.spans.SpanTermQuery;
 
 import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
 
 /**
  * Binder for a single query clause.
  */
 @API(API.Status.UNSTABLE)
 public abstract class LuceneQueryClause implements PlanHashable {
-    public abstract Query bind(@Nonnull FDBRecordStoreBase<?> store, @Nonnull Index index, @Nonnull EvaluationContext context);
+    @Nonnull
+    private final LuceneQueryType queryType;
+
+    protected LuceneQueryClause(@Nonnull final LuceneQueryType queryType) {
+        this.queryType = queryType;
+    }
+
+    @Nonnull
+    public LuceneQueryType getQueryType() {
+        return queryType;
+    }
+
+    public abstract BoundQuery bind(@Nonnull FDBRecordStoreBase<?> store, @Nonnull Index index, @Nonnull EvaluationContext context);
 
     public abstract void getPlannerGraphDetails(@Nonnull ImmutableList.Builder<String> detailsBuilder, @Nonnull ImmutableMap.Builder<String, Attribute> attributeMapBuilder);
+
+    @Nonnull
+    protected BoundQuery toBoundQuery(@Nonnull final Query luceneQuery) {
+        return BoundQuery.ofLuceneQueryWithQueryType(luceneQuery, getQueryType());
+    }
+
+    /**
+     * Parse the Lucene query to get all the mapping from field to terms.
+     * @param query lucene query to extract all terms from
+     */
+    @Nonnull
+    protected static Map<String, Set<String>> getHighlightingTermsMap(@Nonnull final Query query) {
```

Comment:
This is code I took from the highlighting code path and just morphed it into this shape without changing any of its conceptual logic.